mantidproject/mantid
# Make LoadGeometry::isNexus more specific

**Description of work.**



`LoadGeometry::isNexus()` is called by `LoadEventNexus` and `LoadInstrument` to determine if the NeXus files contains geometry information and hence whether to load a Mantid geometry from the file directly or not. However, `LoadGeometry::isNexus()` doesn't actually check if all the fields needed by `NexusGeometry::Parser` actually exists in the file - it just returns `true` if there are _any_ fields with `NXtransformation` or `NX*geometry` attributes in the file. As these are standard NeXus classes which _must_ be used to specify any kind of geometry, many NeXus files will have these fields but may not necessarily specify a valid **Mantid** geometry (that is, a _detector_ geometry Mantid understands).

This is the case with a proposed new modification to the raw data files generated by excitations instruments at ISIS which will describe the instrument up to the sample using NeXus classes in order to allow analysis programs to be able to compute the instrument resolution. Although these files will have fields with `NXtransformation` and `NXoff_geometry` attributes, they will *not* specify detector pixels in Cartesians (they will use polar coordinates as is standard for ISIS raw data files), so they cannot be loaded by `NexusGeometry::Parser`. However, the presence of `NXtransformation` / `NXoff_geometry` tags will cause `LoadGeometry::isNexus` to return `true` and hence for `LoadEventNexus` to try to use `NexusGeometry::Parser` to read the detector geometry instead of using the dedicated ISIS routines (`UpdateInstrumentFromFile::updateFromNeXus`). 

This PR extends the checks in `LoadGeometry::isNexus` to ensure that the NeXus file being checked has all the necessary fields for Mantid geometry. Despite this it **does not** guarantee that files for which `LoadGeometry::isNexus` returns `true` are loadable by Mantid. It only guarantees that files for which `LoadGeometry::isNexus` returns `false` are **not** loadable. To ensure the former, the checks would need to be even more extensive (and expensive...)

**To test:**



Fixes #35256. 


An example of a new data ISIS datafile `MAR28482.nxs` has been uploaded to the test files server. This file should now be loadable by Mantid (previously it was not). Further existing event nexus files should be loadable as before.

*This does not require release notes* because it is an internal change to accommodate a future change in the raw data format for ISIS excitations instrument which should not affect how Mantid currently works.




---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
